### ISSUE TITLE
I fail to implementation this work for semantic segmentation

### ISSUE THREAD

**Issue by 1160742680** [benbergner/cropr#1]
I attempted to reproduce your work using mmseg on the Potsdam dataset, but mIoU is only in the single digits. Here are the modifications I made: I used the ViT+segvit architecture and removed the head used for classification, and replace the original pred with the output of CrossAttention, x_aggr, and concatenated it with x before inputting it into the decoder. There were no other significant changes, and hyperparameters closely related to the model, such as v_proj, were kept consistent with the parameters in the main configuration. I hope you can help me identify where the issue might lie. Thank you very much.
This is the main configuration.
cropr_cfg=dict(
                pruning_rate= 40,
                llf= True,
                use_cropr=True,
                num_queries= 1024,
                num_heads= 6,
                pre_attn_norm=False,
                q_proj=False,
                k_proj=False,
                v_proj=False,
                mlp=True,
                mlp_ratio= 4,
        ),

The code is as follows.

class ViT_cropr(VisionTransformer):
    def __init__(self,
                 num_classes,
                 cropr_cfg,
                 freeze=False,
                 **kwargs,
                 ):
        super(ViT_cropr, self).__init__(
            **kwargs,
        )
        self.num_classes = num_classes
        self.use_cropr = cropr_cfg["use_cropr"]
        self.llf = cropr_cfg["llf"] if self.use_cropr else False
        dpr = [
            x.item()
            for x in torch.linspace(0, kwargs["drop_path_rate"], len(self.layers))
        ]
        dpr[-1] = 0.0 if cropr_cfg["use_cropr"] else dpr[-1]#最后一层通过别的方法实现
        if self.use_cropr:
            n_blk = len(self.layers)
            num_cropr_modules = n_blk - 2 if self.llf else n_blk - 1
            schedule = [cropr_cfg["pruning_rate"]] * num_cropr_modules

            self.cropr = nn.ModuleList(
                [
                    Cropr(
                        pruning_rate=schedule[i],
                        num_queries=cropr_cfg["num_queries"],
                        num_classes=num_classes,
                        embed_dim=self.embed_dims,
                        num_heads=cropr_cfg["num_heads"],
                        pre_attn_norm=cropr_cfg["pre_attn_norm"],
                        q_proj=cropr_cfg["q_proj"],
                        k_proj=cropr_cfg["k_proj"],
                        v_proj=cropr_cfg["v_proj"],
                        mlp=cropr_cfg["mlp"],
                        mlp_ratio=cropr_cfg["mlp_ratio"],
                        training=self.training,
                    )
                    for i in range(num_cropr_modules)
                ]
            )

            num_tokens = (kwargs["img_size"][0] // kwargs["patch_size"]) ** 2
            num_remaining_per_block = [
                num_tokens - sum(schedule[: i + 1]) for i in range(num_cropr_modules)
            ]

    def forward(self, inputs):
        B = inputs.shape[0]

        x, hw_shape = self.patch_embed(inputs)

        # stole cls_tokens impl from Phil Wang, thanks
        cls_tokens = self.cls_token.expand(B, -1, -1)
        x = torch.cat((cls_tokens, x), dim=1)
        x = self._pos_embeding(x, hw_shape, self.pos_embed)
        if not self.with_cls_token:
            # Remove class token for transformer encoder input
            x = x[:, 1:]

        hw = hw_shape[0] * hw_shape[1]

        outs = []
        prnd = []
        x_aggrs=[]
        for i, layer in enumerate(self.layers[:-2]):
            x=layer(x)
            if self.training:
                x, x_p,x_aggr = self.cropr[i](x, inference=not self.training)#x_p为剪枝部分
                prnd.append(x_p)
                x_aggrs.append(x_aggr)
            else:
                x, x_p, x_aggr = self.cropr[i](x, inference=not self.training)  # x_p为剪枝部分
                prnd.append(x_p)

        # no pruning applied to penultimate block because pruned tokens would be concatenated next anyways
        x = self.layers[-2](x)
        x = torch.cat([x] + prnd, dim=1)
        # LLF
        x = self.layers[-1](x)
        if self.final_norm:
            x = self.norm1(x)
        if self.with_cls_token:
            # Remove class token and reshape token for decoder head
            out = x[:, 1:]
        else:
            out = x

        B, _, C = out.shape

        out = out.reshape(B, hw_shape[0], hw_shape[1],
                          C).permute(0, 3, 1, 2).contiguous()
        if self.output_cls_token:
            out = [out, x[:, 0]]
        outs.append(out)
        if self.training:
            outs = outs + x_aggrs
        else:
            outs=outs
        return tuple(outs)

**Comment by benbergner** [comment]
Thanks for trying Cropr with another dataset and architecture.

I have no experience with the Potsdam dataset, but on first glance it looks like the dataset is very small (38 images à 6k x 6k resolution), which might be too small of a dataset in general.

However, given that the number of queries is 1024 according to your code snippet, I assume you train with windows of size 512 x 512?

A few other things that might be worth to check:
- You set the number of heads to 6. This should be 1 only since all projection layers (qkv + head proj) are identities.
- Have you checked that the unpruned baseline gives expected results?
- Have you tried a less aggressive pruning schedule (e.g., prune just 1 token per Cropr module)? The results should then be very close to the unpruned baseline.
- Do the auxiliary losses decrease during training? Losses for Cropr modules applied to deeper layers should decrease faster.
- Gradients from auxiliary heads should not flow back through the encoder

I still plan to release code for Segmenter + ADE20k once I find the time. This might give you further hints to make it work.

**Comment by 1160742680** [comment]
The results are as anticipated when excluding the Class Cropr . However, once incorporating the Class Cropr , I've encountered subpar outcomes, even with a pruning rate of 0 tokens. This leads me to suspect that there might be some errors within the 
 Class Cropr.
Here is the code, which is largely identical to the provided code. Since I opted not to train with the cls_token, I have removed the line  'scores[:, 0] = math.inf  # retain CLS token' . Additionally, I have adjusted the num_heads parameter to 1. I would greatly appreciate your further assistance. Thank you very much for considering my request.

class CrossAttention(nn.Module):
    """Cross-attention layer with learnable queries."""

    def __init__(
        self,
        embed_dim: int = 1024,
        num_heads: int = 1,
        mlp_ratio: float = 4.0,
        qkv_bias: bool = False,
        pre_attn_norm: bool = False,
        q_proj: bool = False,
        k_proj: bool = False,
        v_proj: bool = False,
        mlp: bool = True,
        num_queries: int = 1,
        training: bool = True,
    ):
        super().__init__()
        assert embed_dim % num_heads == 0, "embed_dim must be divisible by num_heads"

        self.num_queries = num_queries
        self.head_dim = embed_dim // num_heads
        self.scale = self.head_dim**-0.5

        norm_layer = partial(nn.LayerNorm, eps=1e-6)
        self.queries = nn.Parameter(torch.empty(1, num_queries, embed_dim))

        self.q = (
            nn.Linear(embed_dim, embed_dim, bias=qkv_bias) if q_proj else nn.Identity()
        )
        self.k = (
            nn.Linear(embed_dim, embed_dim, bias=qkv_bias) if k_proj else nn.Identity()
        )
        self.attn_norm = norm_layer(embed_dim) if pre_attn_norm else nn.Identity()

        if training:
            self.v = (
                nn.Linear(embed_dim, embed_dim, bias=qkv_bias)
                if v_proj
                else nn.Identity()
            )
            self.proj = (
                nn.Linear(embed_dim, embed_dim) if num_heads > 1 else nn.Identity()
            )

            if mlp:
                self.mlp_norm = norm_layer(embed_dim)
                self.mlp = Mlp(embed_dim, int(embed_dim * mlp_ratio))

        self.init_weights()

    def init_weights(self):
        D = self.queries.shape[-1]
        trunc_normal_tf_(self.queries, std=D**-0.5)

    def forward_scorer(self, x):
        B, N = x.shape[:2]

        x = self.attn_norm(x)

        q_in = self.queries.expand(B, -1, -1)
        q = self.q(q_in).reshape(B, self.num_queries, -1, self.head_dim).transpose(1, 2)
        k = self.k(x).reshape(B, N, -1, self.head_dim).transpose(1, 2)
        attn = q @ k.transpose(-2, -1)


        scores = attn.sum((1, 2))
        return scores

    def forward(self, x):
        B, N, D = x.shape

        x = self.attn_norm(x)

        q_in = self.queries.expand(B, -1, -1)
        M = q_in.shape[1]
        q = self.q(q_in).reshape(B, M, -1, self.head_dim).transpose(1, 2)
        k = self.k(x).reshape(B, N, -1, self.head_dim).transpose(1, 2)
        v = self.v(x).reshape(B, N, -1, self.head_dim).transpose(1, 2)

        attn = q @ k.transpose(-2, -1)  # B, H, M, N
        attn_sm = (attn * self.scale).softmax(dim=-1)
        x = attn_sm @ v

        x = x.transpose(1, 2).reshape(B, M, D).squeeze(1)
        x = self.proj(x)

        if hasattr(self, "mlp"):
            x = x + self.mlp(self.mlp_norm(x))

        scores = attn.sum((1, 2))  # B, N

        return x, scores

class ClassificationHead(nn.Module):
    def __init__(self, embed_dim, num_classes):
        super().__init__()
        self.norm = nn.LayerNorm(embed_dim)
        self.head = nn.Linear(embed_dim, num_classes)

        self.init_weights()

    def init_weights(self):
        trunc_normal_(self.head.weight, std=0.02)
        self.head.bias.data.zero_()

    def forward(self, x):
        x = self.norm(x)
        x = self.head(x)
        return x

class Cropr(nn.Module):
    def __init__(
        self,
        pruning_rate: int = 8,
        num_queries: int = 1,
        num_classes: int = 1000,
        embed_dim: int = 1024,
        num_heads: int = 16,
        pre_attn_norm: bool = True,
        q_proj: bool = True,
        k_proj: bool = True,
        v_proj: bool = True,
        mlp: bool = True,
        mlp_ratio: float = 4.0,
        training: bool = True,
    ):
        super().__init__()
        assert embed_dim % num_heads == 0, "embed_dim must be divisible by num_heads"

        self.pruning_rate = pruning_rate
        self.embed_dim = embed_dim
        self.num_queries = num_queries
        self.num_heads = num_heads
        self.head_dim = embed_dim // num_heads
        self.scale = self.head_dim**-0.5
        self.mlp = mlp

        self.cross_attn = CrossAttention(
            embed_dim=embed_dim,
            num_heads=num_heads,
            mlp_ratio=mlp_ratio,
            pre_attn_norm=pre_attn_norm,
            q_proj=q_proj,
            k_proj=k_proj,
            v_proj=v_proj,
            mlp=mlp,
            num_queries=num_queries,
            training=training,
        )
        if training:
            self.head = ClassificationHead(embed_dim, num_classes)

    def prune(self, x, scores):

        M = x.shape[1]
        num_keep = M - self.pruning_rate

        idx = torch.argsort(scores, dim=1, descending=True, stable=False)
        x_reorder = torch.gather(x, 1, idx.unsqueeze(-1).expand(-1, -1, x.shape[-1]))

        x = x_reorder[:, :num_keep]#keep
        x_r = x_reorder[:, num_keep:]#prune

        return x, x_r
    def forward(self, x, inference=False):
        if inference:
            scores = self.cross_attn.forward_scorer(x)
        else:
            # use detach to stop gradients from flowing back
            x_aggr, scores = self.cross_attn(x.detach())


        x, x_r = self.prune(x, scores)
        if inference:
            return x, x_r,None

        return x, x_r,x_aggr

**Comment by benbergner** [comment]
The code should work as far as image classification is concerned!

The public codebase does not yet support segmentation. In particular, minor additions such as preserving the original patch order and label downsampling are currently missing. These will be released as soon as possible. Please stay tuned :-)

**Comment by 1160742680** [comment]
Thank you very much for your help, and I am looking forward to your further results.

**Comment by benbergner** [comment]
I added code for semantic segmentation to `cropr/segm`.
